core/builder.go

```diff
@@ -62,53 +86,67 @@ func defaultRepo(dstore ds.ThreadSafeDatastore) (repo.Repo, error) {
 	}, nil
 }
 
-func (nb *NodeBuilder) Online() *NodeBuilder {
-	nb.online = true
-	return nb
-}
+func NewNode(ctx context.Context, cfg *BuildCfg) (*IpfsNode, error) {
+	if cfg == nil {
+		cfg = new(BuildCfg)
+	}
 
-func (nb *NodeBuilder) Offline() *NodeBuilder {
-	nb.online = false
-	return nb
-}
+	err := cfg.fillDefaults()
+	if err != nil {
+		return nil, err
+	}
 
-func (nb *NodeBuilder) SetRouting(ro RoutingOption) *NodeBuilder {
-	nb.routing = ro
-	return nb
-}
+	n := &IpfsNode{
+		mode:      offlineMode,
+		Repo:      cfg.Repo,
+		ctx:       ctx,
+		Peerstore: peer.NewPeerstore(),
+	}
+	if cfg.Online {
+		n.mode = onlineMode
+	}
 
-func (nb *NodeBuilder) SetHost(ho HostOption) *NodeBuilder {
-	nb.peerhost = ho
-	return nb
-}
+	// TODO: this is a weird circular-ish dependency, rework it
+	n.proc = goprocessctx.WithContextAndTeardown(ctx, n.teardown)
 
-func (nb *NodeBuilder) SetRepo(r repo.Repo) *NodeBuilder {
-	nb.repo = r
-	return nb
-}
+	success := false
+	defer func() {
+		if !success {
+			n.teardown()
```

Comment:
Note:  call `n.Close()`, or even `n.proc.Close()`. do not call `n.teardown()` directly. or it may be called twice and break semantics (maybe panic)